### PR TITLE
First draft to support stacked listeners

### LINKED ISSUE #56: Stacked traces should not be deactivated
Hey guys :)
More of a question then an actual issue...

I introduced Tracer in our projects and wanted to introduce transaction Id stacking (flow id stacking in Zalando speech ;) ). Unfortunately Tracer does not support this in the current version.

My Intention is to have multiple Traces of the same name running at the same time. Basically I get a call through HTTP with an existing trace. To document where things are going from now on I add a new one to the same name (now there are two active) to make tracing of this specific part of the overall transaction easier.

Is there an easier way of doing this?
Or is this a possible feature for Tracer?

### PR DESCRIPTION
**Pull request by whiskeysierra**
Fixes #56

### CODE REVIEW

File: tracer-core/src/main/java/org/zalando/tracer/StackedTraceListener.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package org.zalando.tracer;
+
+import org.zalando.tracer.TracerFactory.Builder;
```
Comment (JonasJurczok) [inline tracer-core/src/main/java/org/zalando/tracer/StackedTraceListener.java]: unused import?

File: tracer-core/src/main/java/org/zalando/tracer/StackedTraceListener.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * {@link StackedTraceListener} are a special kind of {@link TraceListener}, because they are aware of
+ * {@link Builder#stacked() trace stacking}. A stacked trace will, in contract to {@link TraceListener}, be observed
```
Comment (JonasJurczok) [inline tracer-core/src/main/java/org/zalando/tracer/StackedTraceListener.java]: typo: in contrast to

File: tracer-core/src/main/java/org/zalando/tracer/StackedTracer.java [hunk 3]
```diff
@@ -1,28 +1,38 @@
...
     StackedTracer(final Map<String, Generator> generators,
-            final TraceListener listeners) {
+            final Collection<TraceListener> listeners) {
         this.traces = generators.keySet().stream()
                 .collect(toMap(identity(), name -> ThreadLocal.withInitial(LinkedList::new)));
         this.generators = generators;
-        this.listeners = listeners;
+
+        final Map<Boolean, List<TraceListener>> partitions = listeners.stream()
+                .collect(partitioningBy(StackedTraceListener.class::isInstance));
```
Comment (JonasJurczok) [inline tracer-core/src/main/java/org/zalando/tracer/StackedTracer.java]: You sir are crazy :D
Nice.

**Comment by JonasJurczok** [comment]
Looks good to me :)

File: tracer-core/src/main/java/org/zalando/tracer/StackedTraceListener.java (hunk 1 above)
Comment (whiskeysierra) [inline tracer-core/src/main/java/org/zalando/tracer/StackedTraceListener.java]: It's used as a link in the javadoc comment

File: tracer-core/src/main/java/org/zalando/tracer/StackedTraceListener.java (hunk 2 above)
Comment (whiskeysierra) [inline tracer-core/src/main/java/org/zalando/tracer/StackedTraceListener.java]: Fixed

File: tracer-core/src/main/java/org/zalando/tracer/StackedTracer.java (hunk 3 above)
Comment (whiskeysierra) [inline tracer-core/src/main/java/org/zalando/tracer/StackedTracer.java]: This is how `partitioningBy` is meant to be used, imho.

File: tracer-core/src/main/java/org/zalando/tracer/StackedTraceListener.java (hunk 1 above)
Comment (JonasJurczok) [inline tracer-core/src/main/java/org/zalando/tracer/StackedTraceListener.java]: Ah ok. I didn't check that.. :)

**Comment by JonasJurczok** [comment]
+1

**Pull request merged**